src/main/java/org/dependencytrack/search/IndexManager.java

```diff
@@ -339,10 +396,106 @@ public static void delete(final IndexType indexType) {
+        });
+    }
+
+    /**
+     * Check that the index exists and is not corrupted
      */
-    public static boolean exists(final IndexType indexType) {
-        return getIndexDirectory(indexType).exists();
+    private static boolean isIndexHealthy(final IndexType indexType) {
+        LOGGER.info("Checking the health of index "+indexType.name());
+        File indexDirectoryFile = getIndexDirectory(indexType);
+        LOGGER.debug("Checking FS directory "+indexDirectoryFile.toPath());
+        if(!indexDirectoryFile.exists()) {
+            LOGGER.warn("The index "+indexType.name()+" does not exist");
+            return false;
+        }
+        LOGGER.debug("Checking lucene index health");
+        Directory luceneIndexDirectory = null;
+        CheckIndex checkIndex = null;
+        try {
+            File writeLock = Path.of(indexDirectoryFile.getAbsolutePath(), IndexWriter.WRITE_LOCK_NAME).toFile();
+            if (writeLock.exists()) {
+                LOGGER.debug("Stale lock file detected. Deleting it");
+                writeLock.delete();
+            }
+            luceneIndexDirectory = FSDirectory.open(indexDirectoryFile.toPath());
+            checkIndex = new CheckIndex(luceneIndexDirectory);
+            checkIndex.setFailFast(true);
+            checkIndex.setInfoStream(System.out);
+            CheckIndex.Status status = checkIndex.checkIndex();
+            if(status.clean) {
+                LOGGER.info("The index "+indexType.name()+" is healthy");
+            } else {
+                LOGGER.error("The index " + indexType.name().toLowerCase() + " seems to be corrupted");
+            }
+            return status.clean;
+        } catch (IOException | CheckIndex.CheckIndexException e) {
+            LOGGER.error("The index " + indexType.name().toLowerCase() + " seems to be corrupted", e);
+            return false;
+        } finally {
+            try {
+                if (luceneIndexDirectory != null) {
+                    luceneIndexDirectory.close();
+                }
+                if (checkIndex != null) {
+                    checkIndex.close();
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    /**
+     * Check that the index exists, is not corrupted and is consistent with the database.
+     */
+    public static void checkIndexesConsistency() {
+        Arrays.stream(IndexType.values()).forEach(indexType -> {
+            try (QueryManager qm = new QueryManager()) {
+                LOGGER.info("Checking the index " + indexType.name().toLowerCase());
+                if (!isIndexHealthy(indexType)) {
+                    LOGGER.info("(Re)Building index "+indexType.name().toLowerCase());
+                    LOGGER.debug("Dispatching event to reindex "+indexType.name().toLowerCase());
+                    Event.dispatch(new IndexEvent(IndexEvent.Action.REINDEX, indexType.getClazz()));
+                }
```

Comment:
There should be a `return` here. If the index is unhealthy and a reindex has already been kicked off, there's no point in checking data consistency.

Reply:
Done